TestProjects/ShaderFoundry/Assets/CommonAssets/Editor/EditorTests/PropertyTests/VirtualTexturePropertyTests.cs

```diff
@@ -0,0 +1,276 @@
+using System.Collections.Generic;
+using System.Linq;
+using NUnit.Framework;
+using UnityEngine;
+using static UnityEditor.ShaderFoundry.VirtualTextureLayerAttribute;
+using VirtualTextureLayerData = UnityEditor.ShaderFoundry.UnitTests.VirtualTexturePropertyBlockBuilder.LayerData;
+
+namespace UnityEditor.ShaderFoundry.UnitTests
+{
+    class VirtualTexturePropertyTests : BlockTestRenderer
+    {
+        VirtualTexturePropertyBlockBuilder BuildWithNoOverrides(int layerCount, int layerToSample)
+        {
+            var propBuilder = new VirtualTexturePropertyBlockBuilder
+            {
+                PropertyAttribute = new PropertyAttributeData(),
+                LayerCount = layerCount,
+                LayerToSample = layerToSample,
+            };
+            return propBuilder;
+        }
+
+        static string BuildTextureDefaultValueString(string textureName)
+        {
+            return $"\"{textureName}\" {{}}";
+        }
+
+        static string GetTextureDefaultName(VirtualTexturePropertyBlockBuilder propBuilder, int layerIndex)
+        {
+            var defaultValueExpression = propBuilder.GetTextureDefault(layerIndex);
+            if (defaultValueExpression == null)
+                return null;
+
+            // Parse the texture name from within the quotes
+            var start = defaultValueExpression.IndexOf('"');
+            var end = defaultValueExpression.IndexOf('"', start + 1);
+            return defaultValueExpression.Substring(start + 1, end - start - 1);
+        }
+
+        static List<Color> BuildDefaultColors(int layerCount)
+        {
+            // Just create some layer values that are spread out
+            var layerColors = new List<Color>();
+            for (var i = 0; i < layerCount; ++i)
+                layerColors.Add(new Color(i, i + 0.25f, i + 0.5f, i + 0.75f));
+            return layerColors;
+        }
+
+        static List<Texture2D> BuildTexturesFromColors(IEnumerable<Color> colors)
+        {
+            var textures = new List<Texture2D>();
+            foreach (var color in colors)
+            {
+                var texture = new Texture2D(1, 1, TextureFormat.RGBAFloat, false);
+                texture.SetPixel(0, 0, color);
+                texture.Apply();
+                textures.Add(texture);
+            }
+            return textures;
+        }
+
+        void DestroyTextures(IEnumerable<Texture2D> textures)
+        {
+            foreach (var texture in textures)
+                UnityEngine.Object.DestroyImmediate(texture);
+        }
+
+        void SetupMaterial(Material material, int layerCount, VirtualTexturePropertyBlockBuilder propBuilder, List<Texture2D> textures)
+        {
+            for (var i = 0; i < layerCount; ++i)
+            {
+                var uniformName = propBuilder.GetUniformName(i);
+                material.SetTexture(uniformName, textures[i]);
+            }
+        }
+
+        [TestCase(2, 0)]
```

Comment:
The tests I see here are mostly for defaults or correct input.
I'd like to see more incorrect input tests - accessing non-existent layers and such.

Reply:
Good point, I should add some failure case tests.

Reply:
Ok, I added a bunch of failure cases. The builder helpers were really meant for ease of creating the things which meant it was hard to create bad attributes (e.g. a string instead of an int). I think I achieved this in a reasonable way.

Independently, I was thinking I could unit test the attribute parsing classes (`VirtualTextureAttribute`) but I feel like an end-to-end error check might still be good. Let me know what you think.

Reply:
Thank you!

I think both are valuable.